README.md
## Academy Color Encoding System Developer Resources ##

The Academy Color Encoding System (ACES) is a set of components that facilitates a wide range of motion picture and television workflows while eliminating the ambiguity of legacy file formats.  The system is designed to support both all-digital and hybrid film-digital motion picture workflows.

The basic ACES components are:

* Color encoding and metric specifications, file format specifications, color
transformations, and an open source reference implementation 
* A set of reference images and calibration targets for film scanners and recorders 
* Documentation on the system and software tools

This toolkit is intended to serve as a distribution mechanism for key components of the system, including the reference implementation transforms, reference images, and documentation.

### Package Contents ###
 
* [`documents/`](./documents) – ACES-related documents 
* [`images/`](./images) - "golden" images created using the reference implementation transforms
* [`transforms/`](./transforms) - ACES reference implementation transforms

### Changes from Previous Release ###

Though the "master" branch is 1.1, the current major version of ACES remains 1.  This means the 1.1 update adds a number of transforms but does not change the look or modify the existing core transforms (beyond addressing reported bugs and/or inconsequential formatting/whitespace changes).

As always, you should check the hotfixes and dev branches for the latest bug fixes and new features that will ultimately be rolled into a future version of ACES.  Improvements will continue to be staged on the dev branch for testing as they become available.

Included in ACES 1.1:

* New Features: 
    * Add P3 ODTs:
        * P3D65 (and inverse)
        * P3D65 "D60 simulation" (i.e. D60 adapted white point) (and inverse)
        * P3DCI "D65 simulation" (i.e. D65 adapted white point) (and inverse)
        * P3D65 limited to Rec.709 (inverse not required)
    * Add Rec.2020 ODTs:
        * Rec.2020 limited to Rec.709 (inverse not required)
        * Rec.2020 limited to P3D65 (inverse not required)
    * Add DCDM ODT:
        * DCDM with D65 adapted white point and limited to P3D65 (and inverse)
    * Add additional ACESlib and Utilities functions:
        * SSTS: code for the Single Stage Tone Scale used in HDR Output Transforms
        * OutputTransform: beginning of modules needed for parameterizing Output Transforms
    * Add HDR Output Transforms (RRT+ODT) for Dolby Cinema, Rec.2020 ST-2084 (1000, 2000 and 4000 nit) and HLG (1000 nit).
        * P3D65 (108 cd/m^2) ST.2084 - designed for use in Dolby Cinema (and inverse)
        * Rec.2020 (1000 cd/m^2) ST.2084 (and inverse)
        * Rec.2020 (2000 cd/m^2) ST.2084 (and inverse)
        * Rec.2020 (4000 cd/m^2) ST.2084 (and inverse)
        * Rec.2020 (1000 cd/m^2) HLG (and inverse)
    * Add LMT that can help correct highlight clipping artifacts
    * Add new reference images for new transforms
    * Remove HDR ODTs (and inverses)
* New Documentation:
	* TB-2018-001 - Derivation of the ACES White Point CIE Chromaticity Coordinates
* Bug Fixes:
	* Arri IDT - Improve linearization of LogC data
	* Improve consistency of color space chromaticities in ACESlib modules
    * Update copy and paste typo in ACESproxy document
    * Update ODT functions legal range input variable usage to avoid a situation where it may not execute as intended.
    * Update miscellaneous to local variables in utility functions to avoid clashes with existing global variables
    * Update miscellaneous minor errors in Transform IDs
    * Update miscellaneous transforms missing ACESuserName Tags
* Other:
    * Rename some existing transforms for clarity
    	* Rename DCDM_P3D60 to DCDM_P3D60limited
        * Rename P3DCI to P3DCI_D60sim
        * Rename RGBmonitor to sRGB_100nits
    * Miscellaneous white space fixes in CTL transforms
    * Miscellaneous typo fixes in CTL transform comments
    * Miscellaneous README and CTL comment updates
    * Miscellaneous LaTeX documentation typo and code fixes
    * Update tranformIDs where appropriate
    * Update README and CHANGELOG

For a more detailed list of changes see the [CHANGELOG](./CHANGELOG.md) and in the [commit history](https://github.com/ampas/aces-dev/commits/master).

#### Notes on New ODTs ####

A series of new standard dynamic range (SDR) ODTs are included in this release.  These ODTs were added at the request of the ACES Community based on their production needs.  Examples include: P3 ODTs for devices with a D65 calibration white point, simulation of a D65 white point on a P3 device with a DCI calibration white point, and limiting of output image colorimetry to Rec.709 when using a P3D65 calibrated device.  Additional limiting ODTs were also added for Rec.2020 to restrict the image colorimetry to Rec.709 and P3.  A DCDM ODT with limiting to P3D65 were added to compliment the existing DCDM ODT with limiting to P3D60.  These transforms provide support for additional use cases not included in previous ACES releases.

#### Notes on New HDR Output Transforms ####

ACES 1.1 also includes the first release of a series of Output Transforms that combine the RRT and an ODT into a single transform.  The new Output Transforms replace the previous HDR ODTs.  The new Output Transforms are based on a unified, parametric output function.  These individual Output Transforms pass a series of parameters to the underlying output function to improve the consistency of the image processing operations.  Examples of the parameters that that are specified in the Output Transforms include: display primaries, display white point, display max luminance, display min luminance, luminance reproduction of mid-gray, limiting primaries (if any), surround, display EOTF, etc.  In the future, this will make it trivial to generate Output Transforms for non-standard devices.  

Output Transforms using the underlying parametric output function are only provided for HDR devices in dark surround environments at this time, although the transforms are suitable for use in both dark and dim surround environments (they’ve been used this way on several major motion pictures with a corresponding “trim pass” to adjust contrast and saturation). Modifying the surround parameter inside the individual HDR Output Transforms has no effect when the EOTF is set to either ST-2084(PQ) or HLG.  A functional dim surround compensation for HDR devices will be addressed in a future ACES release.  
   
#### Notes on rename of some existing transforms ####
The CTL file names and values of <ACESuserName> of some of the existing ACES transforms have been modified for clarity.  In each case, the output code values from the transform remains the same.  Academy S-2014-002 states “Any transform updates that do not change the output of that transform shall not require the Transform Identifier to be incremented – e.g. whitespace changes, modifications to code comments, etc.”  For this reason, the values <ACEStransformID> have not been changed.  Implementors should take care to update their user interfaces to display the new values of <ACESuserName> where applicable and to alert end-users of the name changes in their product documentation.

### Versioning ###
 
The links to the current and all past versions of the ACES Developer Resources
can be found at [https://github.com/ampas/aces-dev/releases](https://github.com/ampas/aces-dev/releases).  

Source code is version controlled using the [git version control system](http://git-scm.com/) and hosted on GitHub at [https://github.com/ampas/aces-dev/](https://github.com/ampas/aces-dev/).

Individual files now conform to the ACES System Versioning Specification.  Details can be found in the Academy Specification "S-2014-002 - Academy Color Encoding System - Versioning System" included in [`documents/`](./documents)

### Branch Structure ###

__Master Branch__
 
The current release version of ACES can always be found at the HEAD of the master branch.  The master branch contains no intermediate commits and all commits on the master branch are tagged to represent a release of ACES.

__Dev Branch__
 
Intermediate commits between releases will be staged on the dev branch.  Commits staged on the dev branch, but not yet merged into the master, should be considered as "planned for inclusion" in the next release version.  Commits on the dev branch will ultimately be merged into the master branch as part of a future release.

__Hotfixes Branch__

In some instances it may be necessary to create a hotfixes branch.  The hotfixes branch will include important, but not fully tested, fixes for bugs found in a particular release.  Hotfixes should only be implemented by developers if the bug they are intending to correct is encountered in the course of production and is deemed to be a barrier to using a particular ACES release.  Hotfixes, once fully tested, will be merged into the dev branch, and ultimately the master.

## Prerequisites ##

### Color Transformation Language ###

Color Transformation Language (CTL) can be downloaded from
https://github.com/ampas/CTL

## License Terms for Academy Color Encoding System Components ##

Academy Color Encoding System (ACES) software and tools are provided by the
Academy under the following terms and conditions: A worldwide, royalty-free,
non-exclusive right to copy, modify, create derivatives, and use, in source and
binary forms, is hereby granted, subject to acceptance of this license.

Copyright 2018 Academy of Motion Picture Arts and Sciences (A.M.P.A.S.).
Portions contributed by others as indicated. All rights reserved.

Performance of any of the aforementioned acts indicates acceptance to be bound
by the following terms and conditions:

* Copies of source code, in whole or in part, must retain the above copyright
notice, this list of conditions and the Disclaimer of Warranty.

* Use in binary form must retain the above copyright notice, this list of
conditions and the Disclaimer of Warranty in the documentation and/or other
materials provided with the distribution.

* Nothing in this license shall be deemed to grant any rights to trademarks,
copyrights, patents, trade secrets or any other intellectual property of
A.M.P.A.S. or any contributors, except as expressly stated herein.

* Neither the name "A.M.P.A.S." nor the name of any other contributors to this
software may be used to endorse or promote products derivative of or based on
this software without express prior written permission of A.M.P.A.S. or the
contributors, as appropriate.

This license shall be construed pursuant to the laws of the State of
California, and any disputes related thereto shall be subject to the
jurisdiction of the courts therein.

Disclaimer of Warranty: THIS SOFTWARE IS PROVIDED BY A.M.P.A.S. AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
THE IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE, AND
NON-INFRINGEMENT ARE DISCLAIMED. IN NO EVENT SHALL A.M.P.A.S., OR ANY
CONTRIBUTORS OR DISTRIBUTORS, BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, RESITUTIONARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE
OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

WITHOUT LIMITING THE GENERALITY OF THE FOREGOING, THE ACADEMY SPECIFICALLY
DISCLAIMS ANY REPRESENTATIONS OR WARRANTIES WHATSOEVER RELATED TO PATENT OR
OTHER INTELLECTUAL PROPERTY RIGHTS IN THE ACADEMY COLOR ENCODING SYSTEM, OR
APPLICATIONS THEREOF, HELD BY PARTIES OTHER THAN A.M.P.A.S.,WHETHER DISCLOSED OR
UNDISCLOSED.


CHANGELOG.md
**Version 1.1 (June 15, 2018):**

* New Features: 
    * Add P3 ODTs:
        * P3D65 (and inverse)
        * P3D65 "D60 simulation" (i.e. D60 adapted white point) (and inverse)
        * P3DCI "D65 simulation" (i.e. D65 adapted white point) (and inverse)
        * P3D65 limited to Rec.709 (inverse not required)
    * Add Rec.2020 ODTs:
        * Rec.2020 limited to Rec.709 (inverse not required)
        * Rec.2020 limited to P3D65 (inverse not required)
    * Add DCDM ODT:
        * DCDM with D65 adapted white point and limited to P3D65 (and inverse)
    * Add additional ACESlib and Utilities functions:
        * SSTS: code for the Single Stage Tone Scale used in HDR Output Transforms
        * OutputTransform: beginning of modules needed for parameterizing Output Transforms
    * Add HDR Output Transforms (RRT+ODT):
        * P3D65 (108 cd/m^2) ST.2084 - designed for use in Dolby Cinema (and inverse)
        * Rec.2020 (1000 cd/m^2) ST.2084 (and inverse)
        * Rec.2020 (2000 cd/m^2) ST.2084 (and inverse)
        * Rec.2020 (4000 cd/m^2) ST.2084 (and inverse)
        * Rec.2020 (1000 cd/m^2) HLG (and inverse)
    * Add LMT that can help correct highlight clipping artifacts
    * Add new reference images for new transforms
    * Remove HDR ODTs (and inverses)
    	* ODT.Academy.P3D60_ST2084_1000nits.ctl
    	* ODT.Academy.P3D60_ST2084_2000nits.ctl
    	* ODT.Academy.P3D60_ST2084_4000nits.ctl
    	* ODT.Academy.Rec2020_ST2084_1000nits.ctl
    	* InvODT.Academy.P3D60_ST2084_1000nits.ctl
    	* InvODT.Academy.P3D60_ST2084_2000nits.ctl
    	* InvODT.Academy.P3D60_ST2084_4000nits.ctl
    	* InvODT.Academy.Rec2020_ST2084_1000nits.ctl
* New Documentation:
	* TB-2018-001 - Derivation of the ACES White Point CIE Chromaticity Coordinates
		* Python module with reference implementation of TB-2018-001
		* iPython Notebook with calculation of values used in TB-2018-001
* Bug Fixes:
	* Arri IDT - Improve linearization of LogC data
	* Improve consistency of color space chromaticities in ACESlib modules
    * Update copy and paste typo in ACESproxy document
    * Update ODT functions legal range input variable usage to avoid a situation where it may not execute as intended.
    * Update miscellaneous transforms missing ACESuserName Tags
* Other:
    * Rename some existing transforms for clarity:
        * Rename DCDM_P3D60 to DCDM_P3D60limited
        * Rename P3DCI to P3DCI_D60sim
        * Rename RGBmonitor to sRGB_100nits
    * Miscellaneous white space fixes in CTL transforms
    * Miscellaneous typo fixes in CTL transform comments
    * Miscellaneous README and CTL comment updates
    * Miscellaneous LaTeX documentation typo and code fixes
    * Update ACEStranformIDs where appropriate
    * Update README and CHANGELOG

**Version 1.0.3 (September 20, 2016):**

  * New Features: 
     * Add new ACEScct color correction working space transforms
     * Add ACEScct specification document
     * Add Sony S-Log3 / S-Gamut3 IDTs
     * Add functions to convert between premultiplied and straight alpha
     * Add D65 RGB Monitor ODT
     * Add new reference images for new transforms
  * Bug Fixes:
     * Update copy and paste typo in ACESproxy document
     * Update ODT functions legal range input variable usage to avoid a situation where it may not execute as intended.
     * Update miscellaneous to local variables in utility functions to avoid clashes with existing global variables
     * Update miscellaneous minor errors in Transform IDs
     * Update miscellaneous transforms missing ACESuserName Tags
  * Other:
     * Update IDT READMEs to reflect latest manufacturer provided information including broken links
     * Restructure utility and lib functions directories for use clarification
     * Restructure directories to consolidate CSC transforms
     * Update equation variables names in ACEScc and ACESproxy documents for greater clarity 
     * Miscellaneous math simplifications in utility functions
     * Miscellaneous white space fixes in CTL transforms
     * Miscellaneous typo fixes in CTL transform comments
     * Remove version number from CTL file names
     * Add Python script to rename CTL files based on TransformID
     * Update all documents to remove version numbers and use date as unique identifier
     * Update all documents to use vector logo
     * Update reference images to reflect code changes
     * Update README and CHANGELOG

**Version 1.0.2 (April 12, 2016):**

  * Add Missing chromatic adaptation transform in Rec2020 1000-nit InvODT
  * Fix ACEStransformID references in DolbyPQ utility CTLs
  * Fix lingering ST2048 references
  * Fix typo in file names in images/README
  * Fix file names in images README
  * Various minor typographical and stylistic fixes
  * Rename all references to ST2084 from ST2048
  * Tabs => spaces and blank space cleanup
  * Update README and CHANGELOG

**Version 1.0.1 (September 4, 2015):**

  * Add ACES technical documentation as LaTeX source to facilitate tracking of document revisions
  * Add utility functions for making the OCIO config
  * Bug fixes:
      * Fix "ACES to ACESproxy" transforms - Correct a bug in piecewise function logic for lin_2_acesProxy() 
  * Improve readability and robustness of some transforms based on suggestions from users
  * Fix various typos in CTL transform comments and documentation
  * Fix a broken link to the reference images
  * Modify HDR ODT transforms
      * Adjust the lowest four coefficients for the HDR ODT tonescales in conjunction with a small offset to allow for obtaining a code value of 0
      * Add a Rec.2020 version of the 1000 nit ODT
      * Change HDR transforms references to "PQ" to "ST2084"

**Version 1.0 (December 20, 2014):**

  * Additional transforms, encodings, documents, and reference images are included as part of the ACES Version 1.0 release. Please carefully review the ACES Version 1.0 documentation package for details on new features and enhancements for ACES Version 1.0

transforms/python/ssts_sliderPlot.py
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.widgets import Slider, Button
from ssts import ssts, TsPoint
from math import log10, log

def aces2stops( aces):
    return np.log2(aces)-np.log2(0.18)

hgt = 0.04
wdt = 0.23
hspc = 0.1
ledge = 0.05

fig, ax = plt.subplots( figsize=(12,7) )
plt.subplots_adjust(left=ledge, top=0.97, bottom=0.4, right=0.97)

aces = 0.18*pow(2.,np.arange(-20,21,0.5))
Y = ssts(aces)

x = aces2stops(aces)
y = np.log10(Y)
l, = plt.plot( x, y, lw=2, color='red')

minxRange = [-20.,-15.,-6.5]
minyRange = [0.0001,0.0001,0.02]
minsRange = [0.,0.,1.]
midxRange = [-5.,-0.,5.]
midyRange = [0.,4.8,48.]
midsRange = [1.,1.5,2.]
maxxRange = [6.5,15.,20.]
maxyRange = [48.,10000.,10000.]
maxsRange = [0.,0.,1.]
pctLRange = [0.1, 0.25, 0.5]
pctHRange = [0.5, 0.75, 0.9]

minx = 0.18*pow(2.,minxRange[1])
miny = minyRange[1]
midx = 0.18*pow(2.,midxRange[1])
midy = midyRange[1]
maxx = 0.18*pow(2.,maxxRange[1])
maxy = maxyRange[1]
hmin, = plt.plot( aces2stops(minx), log10(miny), color='red', marker='o')
hmid, = plt.plot( aces2stops(midx), log10(midy), color='red', marker='o')
hmax, = plt.plot( aces2stops(maxx), log10(maxy), color='red', marker='o')

hpctL, = plt.plot( midxRange[1]-((midxRange[1]-minxRange[1])/2.), log10(miny)+((log10(midy)-log10(miny))*pctLRange[1]), color='red', marker='x')
hpctH, = plt.plot( midxRange[1]+((maxxRange[1]-midxRange[1])/2.), log10(midy)+((log10(maxy)-log10(midy))*pctHRange[1]), color='red', marker='x')

plt.axis([-20, 20, -4.5, 4.5])
plt.grid(b=True,which='major',axis='both')
plt.xlabel("scene exposure - stops relative to 18% mid-gray")
plt.ylabel("log$_{10}$ luminance ($cd/m^2$)")


axcolor = '#d1d1fa'

ax_minX = plt.axes([ledge, 0.175, wdt, hgt], facecolor='white')
s_minX = Slider(ax_minX, 'X', minxRange[0], minxRange[2], valinit=minxRange[1], valfmt="%1.1f")
ax_minY = plt.axes([ledge, 0.125, wdt, hgt], facecolor='white')
s_minY = Slider(ax_minY, 'Y', minyRange[0], minyRange[2], valinit=minyRange[1], valfmt="%1.4f")
ax_minS = plt.axes([ledge, 0.075, wdt, hgt], facecolor='white')
s_minS = Slider(ax_minS, 'slope', minsRange[0], minsRange[2], valinit=minsRange[1], valfmt="%1.1f")

ax_midX = plt.axes([ledge+wdt+hspc, 0.175, wdt, hgt], facecolor='white')
s_midX = Slider(ax_midX, 'X', midxRange[0], midxRange[2], valinit=midxRange[1], valfmt="%1.1f")
ax_midY = plt.axes([ledge+wdt+hspc, 0.125, wdt, hgt], facecolor='white')
s_midY = Slider(ax_midY, 'Y', midyRange[0], midyRange[2], valinit=midyRange[1], valfmt="%1.1f")
ax_midS = plt.axes([ledge+wdt+hspc, 0.075, wdt, hgt], facecolor='white')
s_midS = Slider(ax_midS, 'slope', midsRange[0], midsRange[2], valinit=midsRange[1], valfmt="%1.2f")

ax_maxX = plt.axes([ledge+2*wdt+2*hspc, 0.175, wdt, hgt], facecolor='white')
s_maxX = Slider(ax_maxX, 'X', maxxRange[0], maxxRange[2], valinit=maxxRange[1], valfmt="%1.1f")
ax_maxY = plt.axes([ledge+2*wdt+2*hspc, 0.125, wdt, hgt], facecolor='white')
s_maxY = Slider(ax_maxY, 'Y', maxyRange[0], maxyRange[2], valinit=maxyRange[1], valfmt="%i")
ax_maxS = plt.axes([ledge+2*wdt+2*hspc, 0.075, wdt, hgt], facecolor='white')
s_maxS = Slider(ax_maxS, 'slope', maxsRange[0], maxsRange[2], valinit=maxsRange[1], valfmt="%1.1f")

ax_pctL = plt.axes([ledge+0.5*wdt+0.6*hspc, 0.25, wdt, hgt], facecolor='white')
s_pctL = Slider(ax_pctL, '% low', pctLRange[0], pctLRange[2], valinit=pctLRange[1], valfmt="%1.2f")
ax_pctH = plt.axes([ledge+1.5*wdt+1.6*hspc, 0.25, wdt, hgt], facecolor='white')
s_pctH = Slider(ax_pctH, '% high', pctHRange[0], pctHRange[2], valinit=pctHRange[1], valfmt="%1.2f")


def update(val):
    minx = 0.18*pow(2.,s_minX.val)
    miny = s_minY.val
    minsl = s_minS.val
    midx = 0.18*pow(2.,s_midX.val)
    midy = s_midY.val
    midsl = s_midS.val
    maxx = 0.18*pow(2.,s_maxX.val)
    maxy = s_maxY.val
    maxsl = s_maxS.val
    pctLow = s_pctL.val
    pctHigh = s_pctH.val

    Min = TsPoint( minx, miny, minsl)
    Mid = TsPoint( midx, midy, midsl)
    Max = TsPoint( maxx, maxy, maxsl)

    l.set_ydata( np.log10( ssts(aces,Min,Mid,Max,pctLow,pctHigh)) )

    hmin.set_xdata( aces2stops(minx))
    hmin.set_ydata( log10(miny))
    hmid.set_xdata( aces2stops(midx))
    hmid.set_ydata( log10(midy))
    hmax.set_xdata( aces2stops(maxx))
    hmax.set_ydata( log10(maxy))
    hpctL.set_xdata( s_midX.val-((s_midX.val-s_minX.val)/2.) )
    hpctL.set_ydata( log10(miny)+((log10(midy)-log10(miny))*pctLow) )
    hpctH.set_xdata( s_midX.val+((s_maxX.val-s_midX.val)/2.) )
    hpctH.set_ydata( log10(midy)+((log10(maxy)-log10(midy))*pctHigh) )
    
    fig.canvas.draw_idle()
s_minX.on_changed(update)
s_minY.on_changed(update)
s_minS.on_changed(update)
s_midX.on_changed(update)
s_midY.on_changed(update)
s_midS.on_changed(update)
s_maxX.on_changed(update)
s_maxY.on_changed(update)
s_maxS.on_changed(update)
s_pctL.on_changed(update)
s_pctH.on_changed(update)

# reset button
resetax = plt.axes([0.8, 0.025, 0.1, 0.03])
button = Button(resetax, 'Reset', color=axcolor, hovercolor='0.975')

def reset(event):
    s_minX.reset()
    s_minY.reset()
    s_minS.reset()
    s_midX.reset()
    s_midY.reset()
    s_midS.reset()
    s_maxX.reset()
    s_maxY.reset()
    s_maxS.reset()
    s_pctL.reset()
    s_pctH.reset()
button.on_clicked(reset)



plt.show()

transforms/python/ssts_sliderPlot_constrained.py
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.widgets import Slider, Button
import ssts
from math import log10, log

def aces2stops( aces):
    return np.log2(aces)-np.log2(0.18)

hgt = 0.04
wdt = 0.17
hspc = 0.15
ledge = 0.05

fig, ax = plt.subplots( figsize=(12,7) )
plt.subplots_adjust(left=ledge, top=0.97, bottom=0.25, right=0.97)

# ACES values to plot
aces = 0.18*pow(2.,np.arange(-23,23,0.1))

# "RRT" settings
rrtMin = ssts.TsPoint( 0.18*pow(2.,-15.), 0.0001, 0.0)
rrtMid = ssts.TsPoint( 0.18, 4.8, 1.5)
rrtMax = ssts.TsPoint( 0.18*pow(2.,18.), 10000., 0.0)
rrt = np.log10( ssts.ssts(aces, rrtMin, rrtMid, rrtMax, ssts.lookup_pctLow(rrtMin.x), ssts.lookup_pctHigh(rrtMax.x)) )

# "48-nit" settings
cinemaMin = ssts.TsPoint( 0.18*pow(2.,-6.5), 0.02, 0.0)
cinemaMid = ssts.TsPoint( 0.18, 4.8, 1.5)
cinemaMax = ssts.TsPoint( 0.18*pow(2.,6.5), 48., 0.0)
cinema = np.log10( ssts.ssts(aces, cinemaMin, cinemaMid, cinemaMax, ssts.lookup_pctLow(cinemaMin.x), ssts.lookup_pctHigh(cinemaMax.x)) )

# "key points"
minyRange = [0.0001,0.001,0.02]
midyRange = [2.,4.8,20.]
maxyRange = [48.,1000.,10000.]
minSlopeRange = [0.0, 0.0, 0.15]
maxSlopeRange = [0.0, 0.0, 0.15]

minsl = minSlopeRange[1]
midsl = 1.5
maxsl = maxSlopeRange[1]

miny = minyRange[1]
minx = ssts.lookup_ACESmin( miny)

midy = 4.8
midx = 0.18

maxy = maxyRange[1]
maxx = ssts.lookup_ACESmax( maxy)

pctLow = ssts.lookup_pctLow( minx)
pctHigh = ssts.lookup_pctHigh( maxx)

# Parameterized tone curve
initialMin = ssts.TsPoint( minx, miny, minsl)
initialMid = ssts.TsPoint( midx, midy, midsl)
initialMax = ssts.TsPoint( maxx, maxy, maxsl)
Y = np.log10( ssts.ssts(aces,initialMin,initialMid,initialMax,ssts.lookup_pctLow(initialMin.x),ssts.lookup_pctHigh(initialMax.x)) )
# d1 = ssts.ssts_d1(aces,initialMin,initialMid,initialMax,ssts.lookup_pctLow(initialMin.x),ssts.lookup_pctHigh(initialMax.x))

# Make the initial plot
x = aces2stops(aces)
ll, = plt.plot( x, rrt, lw=1, color='black', linestyle='--', label='"OCES" (i.e. RRT)')
lll, = plt.plot( x, cinema, lw=1, color='black', linestyle=':', label='48-nit cinema')
l, = plt.plot( x, Y, lw=2, color='red', label='Adjustable')
# lderiv, = plt.plot( x, d1, lw=1, color='red', linestyle=':', label='First derivative')

hmin, = plt.plot( aces2stops(minx), log10(miny), color='red', marker='o')
hmid, = plt.plot( aces2stops(midx), log10(midy), color='red', marker='o')
hmax, = plt.plot( aces2stops(maxx), log10(maxy), color='red', marker='o')

# hpctL, = plt.plot( aces2stops(midx)-((aces2stops(midx)-aces2stops(minx))/2.), log10(miny)+((log10(midy)-log10(miny))*pctLow), color='red', marker='x')
# hpctH, = plt.plot( aces2stops(maxx)-((aces2stops(maxx)-aces2stops(midx))/2.), log10(midy)+((log10(maxy)-log10(midy))*pctHigh), color='red', marker='x')

plt.axis([-20, 20, -4.5, 4.5])
plt.grid(b=True,which='major',axis='both')
plt.xlabel("scene exposure - stops relative to 18% mid-gray")
plt.ylabel("log$_{10}$ luminance ($cd/m^2$)")

plt.legend(loc='upper left')

axcolor = '#d1d1fa'

# Create sliders
ax_minY = plt.axes([ledge+0.07, 0.135, wdt, hgt], facecolor='white')
s_minY = Slider(ax_minY, 'black lum ($cd/m^2$)', np.log10(minyRange[0]), np.log10(minyRange[2]), valinit=np.log10(minyRange[1]), valfmt="%0.4f")
s_minY.valtext.set_text( pow(10., s_minY.val))

ax_midY = plt.axes([ledge+wdt+hspc+0.07, 0.135, wdt, hgt], facecolor='white')
s_midY = Slider(ax_midY, '18% lum ($cd/m^2$)', np.log10(midyRange[0]), np.log10(midyRange[2]), valinit=np.log10(midyRange[1]), valfmt="%0.1f")
s_midY.valtext.set_text( pow(10., s_midY.val))

ax_maxY = plt.axes([ledge+2*wdt+2*hspc+0.07, 0.135, wdt, hgt], facecolor='white')
s_maxY = Slider(ax_maxY, 'white lum ($cd/m^2$)', np.log10(maxyRange[0]), np.log10(maxyRange[2]), valinit=np.log10(maxyRange[1]), valfmt="%i")
s_maxY.valtext.set_text( pow(10., s_maxY.val))

ax_minSlope = plt.axes([ledge+0.07, 0.08, wdt, hgt], facecolor='white')
s_minSlope = Slider(ax_minSlope, 'low slope', minSlopeRange[0], minSlopeRange[2], valinit=minSlopeRange[1], valfmt="%0.1f")
s_minSlope.valtext.set_text( s_minSlope.val)

ax_maxSlope = plt.axes([ledge+2*wdt+2*hspc+0.07, 0.08, wdt, hgt], facecolor='white')
s_maxSlope = Slider(ax_maxSlope, 'high slope', maxSlopeRange[0], maxSlopeRange[2], valinit=maxSlopeRange[1], valfmt="%0.1f")
s_maxSlope.valtext.set_text( s_maxSlope.val)

def update(val):
    midy = 4.8
    midx = 0.18
    miny = pow(10.,s_minY.val)
    minx = ssts.lookup_ACESmin(miny)
    maxy = pow(10.,s_maxY.val)
    maxx = ssts.lookup_ACESmax(maxy)
    pctLow = ssts.lookup_pctLow( minx)
    pctHigh = ssts.lookup_pctHigh( maxx)
    minsl = s_minSlope.val
    maxsl = s_maxSlope.val

    Min = ssts.TsPoint( minx, miny, minsl)
    Mid = ssts.TsPoint( midx, midy, midsl)
    Max = ssts.TsPoint( maxx, maxy, maxsl)

    expShift = ssts.lookup_expShift( pow(10.,s_midY.val) )

    l.set_ydata( np.log10( ssts.ssts(aces,Min,Mid,Max,pctLow,pctHigh)) )
    l.set_xdata( aces2stops( ssts.shift(aces,expShift)))
#     lderiv.set_ydata( ssts.ssts_d1(aces,Min,Mid,Max,pctLow,pctHigh))
#     lderiv.set_xdata( aces2stops( ssts.shift(aces,expShift)))

    hmin.set_xdata( aces2stops(ssts.shift(minx,expShift)))
    hmin.set_ydata( log10(miny))
#     hmid.set_xdata( aces2stops(ssts.shift(midx,expShift)))
    hmid.set_xdata( aces2stops(midx))
    hmid.set_ydata( s_midY.val)
    hmax.set_xdata( aces2stops(ssts.shift(maxx,expShift)))
    hmax.set_ydata( log10(maxy))
#     hpctL.set_xdata( s_midX.val-((s_midX.val-s_minX.val)/2.) )
#     hpctL.set_ydata( log10(miny)+((log10(midy)-log10(miny))*pctLow) )
#     hpctH.set_xdata( s_midX.val+((s_maxX.val-s_midX.val)/2.) )
#     hpctH.set_ydata( log10(midy)+((log10(maxy)-log10(midy))*pctHigh) )
    
    s_minY.valtext.set_text( round(miny,4))
    s_midY.valtext.set_text( round(pow(10.,s_midY.val),1))
    s_maxY.valtext.set_text( round(maxy))
    s_minSlope.valtext.set_text( round(minsl,2))
    s_maxSlope.valtext.set_text( round(maxsl,2))
    
    fig.canvas.draw_idle()

s_minY.on_changed(update)
s_midY.on_changed(update)
s_maxY.on_changed(update)
s_minSlope.on_changed(update)
s_maxSlope.on_changed(update)

# Reset button
resetax = plt.axes([0.8, 0.025, 0.1, 0.03])
button = Button(resetax, 'Reset', color=axcolor, hovercolor='0.975')

def reset(event):
    s_minY.reset()
    s_midY.reset()
    s_maxY.reset()
button.on_clicked(reset)


# Preset buttons
presetaxcolor = '#92CCEA'
presetax1 = plt.axes([0.05, 0.025, 0.1, 0.03])
button1 = Button(presetax1, '"OCES"', color=presetaxcolor)

presetax2 = plt.axes([0.2, 0.025, 0.1, 0.03])
button2 = Button(presetax2, 'x300', color=presetaxcolor)

presetax3 = plt.axes([0.35, 0.025, 0.1, 0.03])
button3 = Button(presetax3, 'PRM-4220', color=presetaxcolor)

presetax4 = plt.axes([0.5, 0.025, 0.1, 0.03])
button4 = Button(presetax4, 'Dolby Cinema', color=presetaxcolor)

presetax5 = plt.axes([0.65, 0.025, 0.1, 0.03])
button5 = Button(presetax5, 'Standard Cinema', color=presetaxcolor)

def preset_OCES(event):
    s_minY.set_val( np.log10(0.0001))
    s_midY.set_val( np.log10(4.8))
    s_maxY.set_val( np.log10(10000.))

def preset_x300(event):
    s_minY.set_val( np.log10(0.0001))
    s_midY.set_val( np.log10(10.0))
    s_maxY.set_val( np.log10(1000.))

def preset_prm4220(event):
    s_minY.set_val( np.log10(0.005))
    s_midY.set_val( np.log10(10.0))
    s_maxY.set_val( np.log10(600.))

def preset_dolbyCinema(event):
    s_minY.set_val( np.log10(0.0001))
    s_midY.set_val( np.log10(7.2))
    s_maxY.set_val( np.log10(108.))

def preset_cinema(event):
    s_minY.set_val( np.log10(0.02))
    s_midY.set_val( np.log10(4.8))
    s_maxY.set_val( np.log10(48.))
    
button1.on_clicked(preset_OCES)
button2.on_clicked(preset_x300)
button3.on_clicked(preset_prm4220)
button4.on_clicked(preset_dolbyCinema)
button5.on_clicked(preset_cinema)


plt.show()